__STDC_VERSION__ regression in master

**Describe the bug**
In v0.12, this all works, but in master, compiling zlib using stack_snapshot with centos7 gives
```
ERROR: /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/external/stackage/BUILD.bazel:12:1: HaskellCabalLibrary @stackage//:zlib failed (Exit 1) cabal_wrapper failed: error executing command bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper lib:zlib zlib-0.6.2.1 false external/stackage/zlib-0.6.2.1/Setup.hs external/stackage/zlib-0.6.2.1 ... (remaining 8 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:29:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/3/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpykxn9foy/tmp/ghc17613_0/ghc_7.c:2:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Stg.h:29:3: error:
     error: #error __STDC_VERSION__ does not advertise C99, C++11 or later
     # error __STDC_VERSION__ does not advertise C99, C++11 or later
       ^
   |
29 | # error __STDC_VERSION__ does not advertise C99, C++11 or later
   |   ^

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:183:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/3/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpykxn9foy/tmp/ghc17613_0/ghc_7.c:2:
../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h: In function ‘overwritingClosure_’:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     error: ‘for’ loop initial declarations are only allowed in C99 mode
         for (uint32_t i = offset; i < size; i++) {
         ^
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     note: use option -std=c99 or -std=gnu99 to compile your code
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^
```

**To Reproduce**
In a new directory
BUILD.bazel is empty
WORKSPACE is (note the commented out options for the `git_repository` line)
```
workspace(name="TestProject")
load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
git_repository(
    name="rules_haskell",
    remote="https://github.com/tweag/rules_haskell.git",
    # commit="ea3f13165dae68d5c0291ef258fb1cd51ea0372d",  # 0.12 works
    commit="57081449c34c0a3de429759837f61a0b70a8f3e2",  # latest master doesn't
)
load("@rules_haskell//haskell:repositories.bzl", "rules_haskell_dependencies")
rules_haskell_dependencies()
load("@rules_haskell//haskell:toolchain.bzl", "rules_haskell_toolchains")
rules_haskell_toolchains(
    version="8.6.5",
    compiler_flags=["-optc-std=c99"],
    locale = "en_US.UTF-8",
)
http_archive(
    name="alex",
    build_file_content="""
load("@rules_haskell//haskell:cabal.bzl", "haskell_cabal_binary")
haskell_cabal_binary(
    name="alex",
    srcs=glob(["**"]),
    visibility=["//visibility:public"],
    compiler_flags=["-optc-std=c99"],
    verbose=False,
)
""",
    strip_prefix="alex-3.2.4",
    urls=["https://hackage.haskell.org/package/alex-3.2.4/alex-3.2.4.tar.gz"],
    sha256="d58e4d708b14ff332a8a8edad4fa8989cb6a9f518a7c6834e96281ac5f8ff232",
)
load("@rules_haskell//haskell:cabal.bzl", "stack_snapshot")
stack_snapshot(
    name="stackage",
    tools=["@alex"],
    snapshot="lts-14.27",
    haddock=False,
    visibility=["//visibility:public"],
    packages=[
        "zlib",
    ]
)
```
I used the following Dockerfile to run centos7
```
FROM centos:7
ENV BAZEL_URL=https://copr.fedorainfracloud.org/coprs/vbatts/bazel/repo/epel-7/vbatts-bazel-epel-7.repo
RUN curl -sSL $BAZEL_URL -o /etc/yum.repos.d/vbatts-bazel-epel-7.repo && \
    yum update -y && \
    yum group install -y "Development Tools" && \
    yum install -y bazel3 python3 gmp-devel ncurses-devel zlib-devel && \
    yum clean all && \
    rm -rf /var/cache/yum
```

**Expected behavior**
Compiling zlib should continue to work in master on centos7. The compiler_flags don't appear to be propagating to the stackage builds. Alternatively, if I need to add -std=c99 to every stack package, there should be an option to do so (there wasn't any as far as I could see).

**Environment**
 - OS name + version: centos 7
 - Bazel version: 3.1.0
 - Version of the rules: v0.12 or commit 57081449c34c0a3de429759837f61a0b70a8f3e2

**Additional context**
If I find time, I will bisect the history to isolate the issue.

Curious, it didn't work for me. `bazel2` => v2.2.0. I also downloaded 2.1.0 on an ec2 instance and I see the same results. Let me try again with the latest master. 
```
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
DEBUG: /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/external/rules_haskell/haskell/private/versions.bzl:60:5: WARNING: bazel version is too recent. Supported versions range from 2.1.0 to 2.1.0, but found: 2.2.0
INFO: Analyzed target @stackage//:zlib (31 packages loaded, 1040 targets configured).
INFO: Found 1 target...
ERROR: /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/external/stackage/BUILD.bazel:12:1: HaskellCabalLibrary @stackage//:zlib failed (Exit 1) cabal_wrapper failed: error executing command bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper lib:zlib zlib-0.6.2.1 false external/stackage/zlib-0.6.2.1/Setup.hs external/stackage/zlib-0.6.2.1 ... (remaining 8 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:29:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpcgjyimbe/tmp/ghc5093_0/ghc_7.c:2:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Stg.h:29:3: error:
     error: #error __STDC_VERSION__ does not advertise C99, C++11 or later
     # error __STDC_VERSION__ does not advertise C99, C++11 or later
       ^
   |
29 | # error __STDC_VERSION__ does not advertise C99, C++11 or later
   |   ^

In file included from ../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/Rts.h:183:0: error:
    0,
                     from /root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpcgjyimbe/tmp/ghc5093_0/ghc_7.c:2:
../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h: In function ‘overwritingClosure_’:

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     error: ‘for’ loop initial declarations are only allowed in C99 mode
         for (uint32_t i = offset; i < size; i++) {
         ^
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^

../../../../../../../../external/rules_haskell_ghc_linux_amd64/lib/include/rts/storage/ClosureMacros.h:558:5: error:
     note: use option -std=c99 or -std=gnu99 to compile your code
    |
558 |     for (uint32_t i = offset; i < size; i++) {
    |     ^
`cc_wrapper-python' failed in phase `C Compiler'. (Exit code: 1)
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper.runfiles/rules_haskell/haskell/cabal_wrapper.py", line 194, in <module>
    run([runghc] + runghc_args + [setup, "build", "--verbose=0", "--builddir=" + distdir])
  File "/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/host/bin/external/rules_haskell/haskell/cabal_wrapper.runfiles/rules_haskell/haskell/cabal_wrapper.py", line 42, in run
    subprocess.run(cmd, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, *args, **kwargs)
  File "/usr/lib64/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/external/rules_haskell_ghc_linux_amd64/bin/runghc', '/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/external/stackage/zlib-0.6.2.1/Setup.hs', 'build', '--verbose=0', '--builddir=/root/.cache/bazel/_bazel_root/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/2/execroot/TestProject/bazel-out/k8-fastbuild/bin/external/stackage/zlib-0.6.2.1/tmpcgjyimbe']' returned non-zero exit status 1.
Target @stackage//:zlib failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 613.118s, Critical Path: 36.61s
INFO: 1 process: 1 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
```

--- a/WORKSPACE
+++ b/WORKSPACE
@@ -166,6 +166,12 @@ test_repl_ghci_args = [
     "-XOverloadedStrings",
 ]
 
+test_cabalopts = [
+    # Used by `tests/cabal-toolchain-flags`
+    "--ghc-option=-DTESTS_TOOLCHAIN_CABALOPTS",
+    "--haddock-option=--optghc=-DTESTS_TOOLCHAIN_CABALOPTS",
+]
+
 load(
     "@rules_haskell//haskell:nixpkgs.bzl",
     "haskell_register_ghc_nixpkgs",
@@ -173,6 +179,7 @@ load(
 
 haskell_register_ghc_nixpkgs(
     attribute_path = "",
+    cabalopts = test_cabalopts,
     compiler_flags = test_compiler_flags,
     haddock_flags = test_haddock_flags,
     locale_archive = "@glibc_locales//:locale-archive",
@@ -188,6 +195,7 @@ load(
 )
 
 haskell_register_ghc_bindists(
+    cabalopts = test_cabalopts,
     compiler_flags = test_compiler_flags,
     version = test_ghc_version,
 )

--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -133,6 +133,7 @@ def _prepare_cabal_inputs(
         setup_deps,
         setup_dep_info,
         srcs,
+        cabalopts,
         compiler_flags,
         flags,
         generate_haddock,
@@ -215,6 +216,7 @@ def _prepare_cabal_inputs(
     if not hs.toolchain.is_darwin and not hs.toolchain.is_windows:
         # See Note [No PIE when linking] in haskell/private/actions/link.bzl
         args.add("--ghc-option=-optl-no-pie")
+    args.add_all(cabalopts)
     args.add_all(compiler_flags, format_each = "--ghc-option=%s")
     if dynamic_binary:
         args.add_all(
@@ -312,6 +314,7 @@ def _haskell_cabal_library_impl(ctx):
     )
     with_profiling = is_profiling_enabled(hs)
 
+    user_cabalopts = _expand_make_variables("cabalopts", ctx, ctx.attr.cabalopts)
     user_compile_flags = _expand_make_variables("compiler_flags", ctx, ctx.attr.compiler_flags)
     cabal = _find_cabal(hs, ctx.files.srcs)
     setup = _find_setup(hs, cabal, ctx.files.srcs)
@@ -382,6 +385,7 @@ def _haskell_cabal_library_impl(ctx):
         setup_deps = setup_deps,
         setup_dep_info = setup_dep_info,
         srcs = ctx.files.srcs,
+        cabalopts = user_cabalopts,
         compiler_flags = user_compile_flags,
         flags = ctx.attr.flags,
         generate_haddock = ctx.attr.haddock,
@@ -514,6 +518,9 @@ haskell_cabal_library = rule(
             aspects = [haskell_cc_libraries_aspect],
             doc = "Dependencies for custom setup Setup.hs.",
         ),
+        "cabalopts": attr.string_list(
+            doc = """Additional flags to pass to `Setup.hs configure`. Subject to make variable expansion.""",
+        ),
         "compiler_flags": attr.string_list(
             doc = """Flags to pass to Haskell compiler, in addition to those defined
             the cabal file. Subject to Make variable substitution.""",
@@ -609,6 +616,7 @@ def _haskell_cabal_binary_impl(ctx):
     posix = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
 
     exe_name = ctx.attr.exe_name if ctx.attr.exe_name else hs.label.name
+    user_cabalopts = _expand_make_variables("cabalopts", ctx, ctx.attr.cabalopts)
     user_compile_flags = _expand_make_variables("compiler_flags", ctx, ctx.attr.compiler_flags)
     cabal = _find_cabal(hs, ctx.files.srcs)
     setup = _find_setup(hs, cabal, ctx.files.srcs)
@@ -644,6 +652,7 @@ def _haskell_cabal_binary_impl(ctx):
         setup_deps = setup_deps,
         setup_dep_info = setup_dep_info,
         srcs = ctx.files.srcs,
+        cabalopts = user_cabalopts,
         compiler_flags = user_compile_flags,
         flags = ctx.attr.flags,
         generate_haddock = False,
@@ -711,6 +720,9 @@ haskell_cabal_binary = rule(
         "setup_deps": attr.label_list(
             aspects = [haskell_cc_libraries_aspect],
             doc = "Dependencies for custom setup Setup.hs.",
+        ),
+        "cabalopts": attr.string_list(
+            doc = """Additional flags to pass to `Setup.hs configure`. Subject to make variable expansion.""",
         ),
         "compiler_flags": attr.string_list(
             doc = """Flags to pass to Haskell compiler, in addition to those defined

--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -215,6 +215,7 @@ def _prepare_cabal_inputs(
     if not hs.toolchain.is_darwin and not hs.toolchain.is_windows:
         # See Note [No PIE when linking] in haskell/private/actions/link.bzl
         args.add("--ghc-option=-optl-no-pie")
+    args.add_all(hs.toolchain.cabalopts)
     args.add_all(cabalopts)
     if dynamic_binary:
         args.add_all(

--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -134,7 +134,6 @@ def _prepare_cabal_inputs(
         setup_dep_info,
         srcs,
         cabalopts,
-        compiler_flags,
         flags,
         generate_haddock,
         cabal_wrapper,
@@ -217,7 +216,6 @@ def _prepare_cabal_inputs(
         # See Note [No PIE when linking] in haskell/private/actions/link.bzl
         args.add("--ghc-option=-optl-no-pie")
     args.add_all(cabalopts)
-    args.add_all(compiler_flags, format_each = "--ghc-option=%s")
     if dynamic_binary:
         args.add_all(
             [
@@ -315,7 +313,12 @@ def _haskell_cabal_library_impl(ctx):
     with_profiling = is_profiling_enabled(hs)
 
     user_cabalopts = _expand_make_variables("cabalopts", ctx, ctx.attr.cabalopts)
-    user_compile_flags = _expand_make_variables("compiler_flags", ctx, ctx.attr.compiler_flags)
+    if ctx.attr.compiler_flags:
+        print("WARNING: compiler_flags attribute is deprecated. Use cabalopts instead.")
+        user_cabalopts.extend([
+            "--ghc-option=" + opt
+            for opt in _expand_make_variables("compiler_flags", ctx, ctx.attr.compiler_flags)
+        ])
     cabal = _find_cabal(hs, ctx.files.srcs)
     setup = _find_setup(hs, cabal, ctx.files.srcs)
     package_database = hs.actions.declare_file(
@@ -386,7 +389,6 @@ def _haskell_cabal_library_impl(ctx):
         setup_dep_info = setup_dep_info,
         srcs = ctx.files.srcs,
         cabalopts = user_cabalopts,
-        compiler_flags = user_compile_flags,
         flags = ctx.attr.flags,
         generate_haddock = ctx.attr.haddock,
         cabal_wrapper = ctx.executable._cabal_wrapper,
@@ -519,11 +521,16 @@ haskell_cabal_library = rule(
             doc = "Dependencies for custom setup Setup.hs.",
         ),
         "cabalopts": attr.string_list(
-            doc = """Additional flags to pass to `Setup.hs configure`. Subject to make variable expansion.""",
+            doc = """Additional flags to pass to `Setup.hs configure`. Subject to make variable expansion.
+
+            Use `--ghc-option=OPT` to configure additional compiler flags.
+            Use `--haddock-option=--optghc=OPT` if these flags are required for haddock generation as well.
+            """,
         ),
         "compiler_flags": attr.string_list(
-            doc = """Flags to pass to Haskell compiler, in addition to those defined
-            the cabal file. Subject to Make variable substitution.""",
+            doc = """DEPRECATED. Use `cabalopts` with `--ghc-option` instead.
+
+            Flags to pass to Haskell compiler, in addition to those defined the cabal file. Subject to Make variable substitution.""",
         ),
         "tools": attr.label_list(
             cfg = "host",
@@ -617,7 +624,12 @@ def _haskell_cabal_binary_impl(ctx):
 
     exe_name = ctx.attr.exe_name if ctx.attr.exe_name else hs.label.name
     user_cabalopts = _expand_make_variables("cabalopts", ctx, ctx.attr.cabalopts)
-    user_compile_flags = _expand_make_variables("compiler_flags", ctx, ctx.attr.compiler_flags)
+    if ctx.attr.compiler_flags:
+        print("WARNING: compiler_flags attribute is deprecated. Use cabalopts instead.")
+        user_cabalopts.extend([
+            "--ghc-option=" + opt
+            for opt in _expand_make_variables("compiler_flags", ctx, ctx.attr.compiler_flags)
+        ])
     cabal = _find_cabal(hs, ctx.files.srcs)
     setup = _find_setup(hs, cabal, ctx.files.srcs)
     package_database = hs.actions.declare_file(
@@ -653,7 +665,6 @@ def _haskell_cabal_binary_impl(ctx):
         setup_dep_info = setup_dep_info,
         srcs = ctx.files.srcs,
         cabalopts = user_cabalopts,
-        compiler_flags = user_compile_flags,
         flags = ctx.attr.flags,
         generate_haddock = False,
         cabal_wrapper = ctx.executable._cabal_wrapper,
@@ -722,11 +733,16 @@ haskell_cabal_binary = rule(
             doc = "Dependencies for custom setup Setup.hs.",
         ),
         "cabalopts": attr.string_list(
-            doc = """Additional flags to pass to `Setup.hs configure`. Subject to make variable expansion.""",
+            doc = """Additional flags to pass to `Setup.hs configure`. Subject to make variable expansion.
+
+            Use `--ghc-option=OPT` to configure additional compiler flags.
+            Use `--haddock-option=--optghc=OPT` if these flags are required for haddock generation as well.
+            """,
         ),
         "compiler_flags": attr.string_list(
-            doc = """Flags to pass to Haskell compiler, in addition to those defined
-            the cabal file. Subject to Make variable substitution.""",
+            doc = """DEPRECATED. Use `cabalopts` with `--ghc-option` instead.
+
+            Flags to pass to Haskell compiler, in addition to those defined the cabal file. Subject to Make variable substitution.""",
         ),
         "tools": attr.label_list(
             cfg = "host",
@@ -1780,7 +1796,7 @@ haskell_cabal_library(
     setup_deps = {setup_deps},
     tools = {tools},
     visibility = {visibility},
-    compiler_flags = ["-w", "-optF=-w"],
+    cabalopts = ["--ghc-option=-w", "--ghc-option=-optF=-w"],
     verbose = {verbose},
     unique_name = True,
 )
@@ -1815,7 +1831,7 @@ haskell_cabal_binary(
     deps = {deps},
     tools = {tools},
     visibility = ["@{workspace}-exe//{name}:__pkg__"],
-    compiler_flags = ["-w", "-optF=-w"],
+    cabalopts = ["--ghc-option=-w", "--ghc-option=-optF=-w"],
     verbose = {verbose},
 )
 """.format(

--- a/haskell/ghc_bindist.bzl
+++ b/haskell/ghc_bindist.bzl
@@ -308,6 +308,7 @@ haskell_toolchain(
     compiler_flags = {compiler_flags},
     haddock_flags = {haddock_flags},
     repl_ghci_args = {repl_ghci_args},
+    cabalopts = {cabalopts},
     visibility = ["//visibility:public"],
     locale = "{locale}",
 )
@@ -318,6 +319,7 @@ haskell_toolchain(
         compiler_flags = ctx.attr.compiler_flags,
         haddock_flags = ctx.attr.haddock_flags,
         repl_ghci_args = ctx.attr.repl_ghci_args,
+        cabalopts = ctx.attr.cabalopts,
         locale = locale,
     )
 
@@ -351,6 +353,7 @@ _ghc_bindist = repository_rule(
         "compiler_flags": attr.string_list(),
         "haddock_flags": attr.string_list(),
         "repl_ghci_args": attr.string_list(),
+        "cabalopts": attr.string_list(),
         "patches": attr.label_list(
             default = [],
             doc =
@@ -418,6 +421,7 @@ def ghc_bindist(
         compiler_flags = None,
         haddock_flags = None,
         repl_ghci_args = None,
+        cabalopts = None,
         locale = None):
     """Create a new repository from binary distributions of GHC.
 
@@ -474,6 +478,7 @@ def ghc_bindist(
         compiler_flags = compiler_flags,
         haddock_flags = haddock_flags,
         repl_ghci_args = repl_ghci_args,
+        cabalopts = cabalopts,
         target = target,
         locale = locale,
         **extra_attrs
@@ -490,6 +495,7 @@ def haskell_register_ghc_bindists(
         compiler_flags = None,
         haddock_flags = None,
         repl_ghci_args = None,
+        cabalopts = None,
         locale = None):
     """Register GHC binary distributions for all platforms as toolchains.
 
@@ -513,6 +519,7 @@ def haskell_register_ghc_bindists(
             compiler_flags = compiler_flags,
             haddock_flags = haddock_flags,
             repl_ghci_args = repl_ghci_args,
+            cabalopts = cabalopts,
             locale = locale,
         )
     local_sh_posix_repo_name = "rules_haskell_sh_posix_local"

--- a/haskell/nixpkgs.bzl
+++ b/haskell/nixpkgs.bzl
@@ -79,6 +79,7 @@ haskell_toolchain(
     is_static = {is_static},
     compiler_flags = {compiler_flags} + {compiler_flags_select},
     haddock_flags = {haddock_flags},
+    cabalopts = {cabalopts},
     repl_ghci_args = {repl_ghci_args},
     # On Darwin we don't need a locale archive. It's a Linux-specific
     # hack in Nixpkgs.
@@ -94,6 +95,7 @@ haskell_toolchain(
             compiler_flags_select = compiler_flags_select,
             haddock_flags = repository_ctx.attr.haddock_flags,
             repl_ghci_args = repository_ctx.attr.repl_ghci_args,
+            cabalopts = repository_ctx.attr.cabalopts,
             locale_archive_arg = "locale_archive = {},".format(repr(locale_archive)) if locale_archive else "",
             locale = repr(repository_ctx.attr.locale),
         ),
@@ -109,6 +111,7 @@ _ghc_nixpkgs_haskell_toolchain = repository_rule(
         "compiler_flags": attr.string_list(),
         "compiler_flags_select": attr.string_list_dict(),
         "haddock_flags": attr.string_list(),
+        "cabalopts": attr.string_list(),
         "repl_ghci_args": attr.string_list(),
         "locale_archive": attr.string(),
         # Unfortunately, repositories cannot depend on each other
@@ -166,6 +169,7 @@ def haskell_register_ghc_nixpkgs(
         compiler_flags_select = None,
         haddock_flags = None,
         repl_ghci_args = None,
+        cabalopts = None,
         locale_archive = None,
         attribute_path = "haskellPackages.ghc",
         sh_posix_attributes = None,
@@ -237,6 +241,7 @@ def haskell_register_ghc_nixpkgs(
         compiler_flags = compiler_flags,
         compiler_flags_select = compiler_flags_select,
         haddock_flags = haddock_flags,
+        cabalopts = cabalopts,
         repl_ghci_args = repl_ghci_args,
         locale_archive = locale_archive,
         locale = locale,

--- a/haskell/toolchain.bzl
+++ b/haskell/toolchain.bzl
@@ -146,6 +146,7 @@ def _haskell_toolchain_impl(ctx):
             compiler_flags = ctx.attr.compiler_flags,
             repl_ghci_args = ctx.attr.repl_ghci_args,
             haddock_flags = ctx.attr.haddock_flags,
+            cabalopts = ctx.attr.cabalopts,
             locale = ctx.attr.locale,
             locale_archive = locale_archive,
             cc_wrapper = struct(
@@ -194,6 +195,15 @@ _haskell_toolchain = rule(
         ),
         "haddock_flags": attr.string_list(
             doc = "A collection of flags that will be passed to haddock.",
+        ),
+        "cabalopts": attr.string_list(
+            doc = """Additional flags to pass to `Setup.hs configure` for all Cabal rules.
+
+            Note, Cabal rules do not read the toolchain attributes `compiler_flags` or `haddock_flags`.
+            Use `--ghc-option=OPT` to configure additional compiler flags.
+            Use `--haddock-option=OPT` to configure additional haddock flags.
+            Use `--haddock-option=--optghc=OPT` if haddock generation requires additional compiler flags.
+            """,
         ),
         "version": attr.string(
             doc = "Version of your GHC compiler. It has to match the version reported by the GHC used by bazel.",
@@ -246,6 +256,7 @@ def haskell_toolchain(
         compiler_flags = [],
         repl_ghci_args = [],
         haddock_flags = [],
+        cabalopts = [],
         locale_archive = None,
         **kwargs):
     """Declare a compiler toolchain.
@@ -291,6 +302,7 @@ def haskell_toolchain(
         compiler_flags = compiler_flags,
         repl_ghci_args = corrected_ghci_args,
         haddock_flags = haddock_flags,
+        cabalopts = cabalopts,
         is_darwin = select({
             "@rules_haskell//haskell/platforms:darwin": True,
             "//conditions:default": False,

--- a/tests/cabal-toolchain-flags/BUILD.bazel
+++ b/tests/cabal-toolchain-flags/BUILD.bazel
@@ -1,0 +1,6 @@
+load("@rules_haskell//haskell:cabal.bzl", "haskell_cabal_binary")
+
+haskell_cabal_binary(
+    name = "cabal-toolchain-flags",
+    srcs = glob(["**"]),
+)

--- a/tests/cabal-toolchain-flags/Main.hs
+++ b/tests/cabal-toolchain-flags/Main.hs
@@ -1,0 +1,10 @@
+{-# LANGUAGE CPP #-}
+
+module Main where
+
+import Data.ByteString
+
+#ifdef TESTS_TOOLCHAIN_CABALOPTS
+main :: IO ()
+main = print ("hello" :: ByteString)
+#endif

--- a/tests/cabal-toolchain-flags/test.cabal
+++ b/tests/cabal-toolchain-flags/test.cabal
@@ -1,0 +1,10 @@
+cabal-version: >=1.10
+name: cabal-toolchain-flags
+version: 0.1.0.0
+build-type: Simple
+
+executable cabal-toolchain-flags
+  build-depends: base, bytestring
+  default-language: Haskell2010
+  extensions: OverloadedStrings
+  main-is: Main.hs

--- a/tests/haskell_cabal_library/BUILD.bazel
+++ b/tests/haskell_cabal_library/BUILD.bazel
@@ -21,6 +21,10 @@ haskell_cabal_library(
         "Lib.hs",
         "lib.cabal",
     ],
+    cabalopts = [
+        "--ghc-option=-DONE=1",
+        "--haddock-option=--optghc=-DONE=1",
+    ],
     flags = [
         "use-base",
         "expose-lib",

--- a/tests/haskell_cabal_library/Lib.hs
+++ b/tests/haskell_cabal_library/Lib.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE ForeignFunctionInterface #-}
 
 module Lib where
@@ -5,4 +6,4 @@ module Lib where
 foreign import ccall "add" add :: Int -> Int -> Int
 
 x :: Int
-x = add 1 1
+x = add ONE ONE

